File: 0001-0500/0030/0030.substring-with-concatenation-of-all-words.cs
/*
 * @lc app=leetcode id=30 lang=csharp
 *
 * [30] Substring with Concatenation of All Words
 */

// @lc code=start
public class Solution {
    public IList<int> FindSubstring(string s, string[] words) {
        IList<int> res = new List<int>();
        if(s == null || s.Length == 0 || words == null || words.Length == 0) return res;
        int wordLen = words[0].Length;
        int wordCount = words.Length;
        int totalLen = wordLen * wordCount;
        if(s.Length < totalLen) return res;
        Dictionary<string, int> wordDict = new Dictionary<string, int>();
        foreach(string word in words) {
            if(wordDict.ContainsKey(word)) {
                wordDict[word]++;
            } else {
                wordDict.Add(word, 1);
            }
        }
        for(int i = 0; i <= s.Length - totalLen; i++) {
            Dictionary<string, int> tempDict = new Dictionary<string, int>(wordDict);
            for(int j = 0; j < wordCount; j++) {
                string word = s.Substring(i + j * wordLen, wordLen);
                if(tempDict.ContainsKey(word)) {
                    if(tempDict[word] == 1) {
                        tempDict.Remove(word);
                    } else {
                        tempDict[word]--;
                    }
                } else {
                    break;
                }
            }
            if(tempDict.Count == 0) {
                res.Add(i);
            }
        }
        return res;
    }
}
// @lc code=end

File: 0001-0500/0030/0030.substring-with-concatenation-of-all-words.cpp
/*
 * @lc app=leetcode id=30 lang=cpp
 *
 * [30] Substring with Concatenation of All Words
 */

// @lc code=start
class Solution
{
public:
    vector<int> findSubstring(string s, vector<string> &words)
    {
        vector<int> res;
        if (words.empty())
            return res;
        int n = s.size(), m = words.size(), len = words[0].size();
        unordered_map<string, int> m1;
        for (auto &a : words)
            ++m1[a];
        for (int i = 0; i < len; ++i)
        {
            unordered_map<string, int> m2;
            int cnt = 0, left = i;
            for (int j = i; j <= n - len; j += len)
            {
                string t = s.substr(j, len);
                if (m1.count(t))
                {
                    ++m2[t];
                    if (m2[t] <= m1[t])
                        ++cnt;
                    else
                    {
                        while (m2[t] > m1[t])
                        {
                            string t1 = s.substr(left, len);
                            --m2[t1];
                            if (m2[t1] < m1[t1])
                                --cnt;
                            left += len;
                        }
                    }
                    if (cnt == m)
                    {
                        res.push_back(left);
                        --m2[s.substr(left, len)];
                        --cnt;
                        left += len;
                    }
                }
                else
                {
                    m2.clear();
                    cnt = 0;
                    left = j + len;
                }
            }
        }
        return res;
    }
};
// @lc code=end
